src/Polytopes/iterators.jl

```diff
@@ -0,0 +1,229 @@
+
+struct HalfSpaceIterator{T} <: AbstractArray{T, 1}
+    A::Polymake.Matrix{Polymake.Rational}
+    b::Polymake.Vector{Polymake.Rational}
+end
+
+Base.IndexStyle(::Type{<:HalfSpaceIterator}) = IndexLinear()
+
+function Base.getindex(iter::HalfSpaceIterator{T}, i::Int64) where T
+    @boundscheck checkbounds(iter.b, i)
+    return T(reshape(iter.A[i, :], 1, :), iter.b[i])
+end
+
+function Base.setindex!(iter::HalfSpaceIterator, val::Halfspace, i::Int64)
+    @boundscheck checkbounds(iter.b, i)
+    iter.A[i, :] = val.a
+    iter.b[i] = val.b
+    return val
+end
+
+Base.firstindex(::HalfSpaceIterator) = 1
+Base.lastindex(iter::HalfSpaceIterator) = length(iter)
+Base.size(iter::HalfSpaceIterator) = (length(iter.b),)
+
+halfspace_matrix_pair(iter::HalfSpaceIterator) = (A = iter.A, b = iter.b)
+
+HalfSpaceIterator(x...) = HalfSpaceIterator{Halfspace}(x...)
+
+###############################
+
+struct PointIterator{T, U} <: AbstractArray{T, 1}
+    m::Polymake.Matrix{U}
+end
+
+Base.IndexStyle(::Type{<:PointIterator}) = IndexLinear()
+
+function Base.getindex(iter::PointIterator{T, U}, i::Int64) where {T, U}
+    @boundscheck checkbounds(iter.m, i, 1)
+    return T{U}(iter.m[i, :])
+end
+
+function Base.setindex!(iter::PointIterator{T}, val::T, i::Int64) where T
+    @boundscheck checkbounds(iter.m, i, 1)
+    iter.m[i, :] = val.p
+    return val
+end
+
+Base.firstindex(::PointIterator) = 1
+Base.lastindex(iter::PointIterator) = length(iter)
+Base.size(iter::PointIterator) = (size(iter.m, 1),)
+
+point_matrix(iter::PointIterator) = iter.m
+
+PointIterator{T, U}(vertices::Union{Oscar.fmpz_mat,AbstractMatrix{Oscar.fmpz}, Oscar.fmpq_mat,AbstractMatrix{Oscar.fmpq}}) where {T, U} = PointIterator{T, U}(matrix_for_polymake(vertices))
+PointIterator{T, U}(vertices::Array{V, 1}) where {T, U, V<:Union{AbstractVector, T}} = PointIterator{T, U}(cat((v' for v in vertices)...; dims = 1))
+
+PointIterator{T}(x...) where T = PointIterator{T, Polymake.Rational}(x...)
+PointIterator(x...) = PointIterator{Points}(x...)
+
+####################
+
+AsTypeIdentitiesP(as::Type{T}) where T<:Union{Points, Polymake.Vector} = Points
+
+AsTypeIdentitiesR(as::Type{T}) where T<:Union{Points} = Ray
+
+AsTypeIdentitiesH(as::Type{T}) where T<:Union{Halfspace, Halfspaces} = Halfspace
+AsTypeIdentitiesH(as::Type{T}) where T<:Union{Polyhedron, Polyhedra} = Polyhedron
+AsTypeIdentitiesH(as::Type{T}) where T<:Pair = Pair{Polymake.Matrix, Polymake.Rational}
+
+AsTypeIdentitiesPoly(as::Type{T}) where T<:Union{Polyhedron, Polyhedra} = Polyhedron
+
+AsTypeIdentitiesC(as::Type{T}) where T<:Union{Cone, Cones} = Cone
+
+AsTypeIdentitiesP(as::Any) = throw(ArgumentError("Unsupported `as` argument :" * string(as)))
+AsTypeIdentitiesH(as::Any) = throw(ArgumentError("Unsupported `as` argument :" * string(as)))
+AsTypeIdentitiesPoly(as::Any) = throw(ArgumentError("Unsupported `as` argument :" * string(as)))
+AsTypeIdentitiesC(as::Any) = throw(ArgumentError("Unsupported `as` argument :" * string(as)))
```

Comment:
What I meant with my earlier comment here is: What do the *names* stand for? What is P / H / C about?

Reply:
These are abbreviations for the context I use them in. `P` stands for `Points`, `C` for `Cone` etc.
Since your last comment I have had some thoughts about this. Luckily, most of it is directly resolved by another advice you have given: letting julia's dispatch take care of the allowed `as::Type{T}` types for methods like `vertices`. That means that we can leave out the different function names and just define
```
asTypeIdentities(::Type{T}) where T<:Union{Ray, Rays} = Ray
asTypeIdentities(::Type{T}) where T<:Union{Halfspace, Halfspaces} = Ray
...
```